upcloud/managed_database.go

```diff
@@ -0,0 +1,693 @@
+	QuerySampleText         string        `json:"query_sample_text"`
+	QuerySampleTimerWait    time.Duration `json:"query_sample_timer_wait"`
+	SchemaName              string        `json:"schema_name"`
+	SumCreatedTmpDiskTables uint64        `json:"sum_created_tmp_disk_tables"`
+	SumCreatedTmpTables     uint64        `json:"sum_created_tmp_tables"`
+	SumErrors               uint64        `json:"sum_errors"`
+	SumLockTime             uint64        `json:"sum_lock_time"`
+	SumNoGoodIndexUsed      uint64        `json:"sum_no_good_index_used"`
+	SumNoIndexUsed          uint64        `json:"sum_no_index_used"`
+	SumRowsAffected         uint64        `json:"sum_rows_affected"`
+	SumRowsExamined         uint64        `json:"sum_rows_examined"`
+	SumRowsSent             uint64        `json:"sum_rows_sent"`
+	SumSelectFullJoin       uint64        `json:"sum_select_full_join"`
+	SumSelectFullRangeJoin  uint64        `json:"sum_select_full_range_join"`
+	SumSelectRange          uint64        `json:"sum_select_range"`
+	SumSelectRangeCheck     uint64        `json:"sum_select_range_check"`
+	SumSelectScan           uint64        `json:"sum_select_scan"`
+	SumSortMergePasses      uint64        `json:"sum_sort_merge_passes"`
+	SumSortRange            uint64        `json:"sum_sort_range"`
+	SumSortRows             uint64        `json:"sum_sort_rows"`
+	SumSortScan             uint64        `json:"sum_sort_scan"`
+	SumTimerWait            time.Duration `json:"sum_timer_wait"`
+	SumWarnings             uint64        `json:"sum_warnings"`
+}
+
+// MarshalJSON implements json.Marshaler for the canonical form of query statistics
+func (m *ManagedDatabaseQueryStatisticsMysql) MarshalJSON() ([]byte, error) {
+	// Duration values are in picoseconds in the canonical form
+	type alias ManagedDatabaseQueryStatisticsMysql
+	out := struct {
+		alias
+		AvgTimerWait         uint64 `json:"avg_timer_wait"`
+		MaxTimerWait         uint64 `json:"max_timer_wait"`
+		MinTimerWait         uint64 `json:"min_timer_wait"`
+		Quantile95           uint64 `json:"quantile_95"`
+		Quantile99           uint64 `json:"quantile_99"`
+		Quantile999          uint64 `json:"quantile_999"`
+		QuerySampleTimerWait uint64 `json:"query_sample_timer_wait"`
+		SumTimerWait         uint64 `json:"sum_timer_wait"`
+	}{
+		alias:                alias(*m),
+		AvgTimerWait:         uint64(m.AvgTimerWait) * 1e3,
+		MaxTimerWait:         uint64(m.MaxTimerWait) * 1e3,
+		MinTimerWait:         uint64(m.MinTimerWait) * 1e3,
+		Quantile95:           uint64(m.Quantile95) * 1e3,
+		Quantile99:           uint64(m.Quantile99) * 1e3,
+		Quantile999:          uint64(m.Quantile999) * 1e3,
+		QuerySampleTimerWait: uint64(m.QuerySampleTimerWait) * 1e3,
+		SumTimerWait:         uint64(m.SumTimerWait) * 1e3,
+	}
+	return json.Marshal(&out)
+}
+
+func (m *ManagedDatabaseQueryStatisticsMysql) UnmarshalJSON(d []byte) error {
+	// Duration values are in picoseconds in the canonical form
```

Comment:
Seems a bit odd that the durations are different across pg (float64 milliseconds?) and mysql (uint64 picoseconds) - could we make them consistent?

Reply:
These are feeded directly from the database engine's statistics tables. I agree with you that we could represent these in unified unit but will need to agree with the team as there's already UI work being done. Should we just represent these as float64 in seconds or would it loose too much precision? Or just return them in nanoseconds so we can directly load them as `time.Duration`?

Reply:
Either option seems fine to me, as long as it's documented. 

I can't see float accuracy being a problem, considering most amount of the significant numbers for our use case should be after the point, but then again, that also points towards just using integers eg. nanoseconds as it seems very unlikely that we'll ever go over the limit of uin64 nanoseconds (eg. 1<<64-1 / 1e9 / 60 / 60 / 24 / 12 =~ 593 years). I guess we could do 'the highest precision we get' (eg. microseconds?) but in case we ever get something that provides higher accuracy, that would be kinda sucky.

So, I guess if I had a gun to my head and was forced to pick, I would go for integer nanoseconds but certainly wouldn't mind float seconds either.

Reply:
I just checked with the team and they are at least happy with nanoseconds normalisation. From the sdk's PoV the limit is the signed int64 and it's plenty. At least it would remove the custom marshaling/unmarshaling.

Reply:
implemented